src/query/graphite/common/aggregation.go

```diff
@@ -61,3 +62,163 @@ func Range(ctx *Context, series ts.SeriesList, renamer SeriesListRenamer) (*ts.S
+	nans := 0
+	sum := 0.0
+	for _, v := range input {
+		if !math.IsNaN(v) {
+			sum += v
+		} else {
+			nans++
+		}
+	}
+	return sum, nans
+}
+
+// SafeAverage returns the average of the input slice and the number of NaNs in the input.
+func SafeAverage(input []float64) (float64, int) {
+	sum, nans := SafeSum(input)
+	count := len(input) - nans
+	return sum / float64(count), nans
+}
+
+// SafeMax returns the maximum value of the input slice and the number of NaNs in the input.
+func SafeMax(input []float64) (float64, int) {
+	nans := 0
+	max := -math.MaxFloat64
+	for _, v := range input {
+		if math.IsNaN(v) {
+			nans++
+			continue
+		}
+		if v > max {
+			max = v
+		}
+	}
+	return max, nans
+}
+
+// SafeMin returns the minimum value of the input slice and the number of NaNs in the input.
+func SafeMin(input []float64) (float64, int) {
+	nans := 0
+	min := math.MaxFloat64
+	for _, v := range input {
+		if math.IsNaN(v) {
+			nans++
+			continue
+		}
+		if v < min {
+			min = v
+		}
+	}
+	return min, nans
+}
+
+// SafeMedian returns the median value of the input slice and the number of NaNs in the input.
+func SafeMedian(input []float64) (float64, int) {
+	safeValues, nans := safeValuesFn(input)
+
+	return ts.Median(safeValues, len(safeValues)), nans
+}
+
+// SafeDiff returns the subtracted value of all the subsequent numbers from the 1st one and the number of NaNs in the input.
+func SafeDiff(input []float64) (float64, int) {
+	safeValues, nans := safeValuesFn(input)
+
+	diff := safeValues[0]
+	for i := 1; i < len(safeValues); i++ {
+		diff = diff - safeValues[i]
+	}
+
+	return diff, nans
+}
+
+// SafeStddev returns the standard deviation value of the input slice and the number of NaNs in the input.
+func SafeStddev(input []float64) (float64, int) {
+	safeAvg, nans := SafeAverage(input)
+
+	sum := 0.0
+	safeValues, _ := safeValuesFn(input)
+	for _, v := range safeValues {
+		sum = sum + (v-safeAvg)*(v-safeAvg)
+	}
+
+	return math.Sqrt(sum / float64(len(safeValues))), nans
+}
+
+// SafeRange returns the range value of the input slice and the number of NaNs in the input.
+func SafeRange(input []float64) (float64, int) {
+	safeMax, nans := SafeMax(input)
+	safeMin, _ := SafeMin(input)
+
+	return safeMax - safeMin, nans
+}
+
+// SafeMul returns the product value of the input slice and the number of NaNs in the input
+func SafeMul(input []float64) (float64, int) {
+	safeValues, nans := safeValuesFn(input)
+
+	product := 1.0
+	for _, v := range safeValues {
+		product = product * v
+	}
+
+	return product, nans
+}
+
+// SafeLast returns the last value of the input slice and the number of NaNs in the input
+func SafeLast(input []float64) (float64, int) {
+	safeValues, nans := safeValuesFn(input)
+	if len(safeValues) >= 1 {
+		return safeValues[len(safeValues)-1], nans
+	} else {
+		return math.NaN(), nans
+	}
+}
+
+func safeValuesFn(input []float64) ([]float64, int) {
+	if len(input) == 0 {
+		return []float64{math.NaN()}, 0
+	}
```

Comment:
Hm, it seems like some code that calls safeValues... checks for this case and some does not.

Perhaps it would be more "go" like and safer if this matched the approach Graphite web does and basically special case the return?
i.e. graphite-web checks for if there was actually values returned after computing safeValues or not:
https://github.com/graphite-project/graphite-web/blob/19f2155a8902f1dc075d4276d384409f1c006764/webapp/graphite/functions/safe.py#L76-L77

So perhaps this function should be something like:
```go
func safeValues(input []float64) ([]float64, int, bool) {
	nans := 0
	safeValues := make([]float64, 0, len(input)) // Zero length but preallocate capacity
	for _, v := range input {
		if !math.IsNaN(v) {
			safeValues = append(safeValues, v)
		} else {
			nans++
		}
	}
	return safeValues, nans, len(safeValues) > 0
}
```

Then you can guarantee most folks will check the resulting `bool` of whether it is safe to use the result or not.

i.e. in SafeStddev you can:
```go
func SafeStddev(input []float64) (float64, int, bool) {
	safeValues, _, ok := safeValues(input)
	if !ok {
		return 0, 0, false
	}

	// Progress normally.
}
```